### ISSUE TITLE
Merge compile error with long long

### ISSUE THREAD

**Issue by sdalton1** [thrust/thrust#658]
Code below doesn't compile using long long and cuda device vectors.
Setting index_type to int or using constant_iterators instead of counting_iterators fixes the issue.

```c++
#include <thrust/device_vector.h>
#include <thrust/host_vector.h>
#include <thrust/merge.h>
#include <thrust/iterator/constant_iterator.h>
#include <thrust/iterator/counting_iterator.h>
#include <thrust/iterator/discard_iterator.h>

int main()
{
    typedef long long index_type;

    thrust::device_vector<index_type> vec(10);

    thrust::merge_by_key(thrust::counting_iterator<index_type>(0), thrust::counting_iterator<index_type>(0),
                         vec.begin(), vec.begin(),
                         thrust::counting_iterator<index_type>(0), thrust::counting_iterator<index_type>(0),
                         thrust::make_discard_iterator(),
                         thrust::make_discard_iterator());

    return 0;
}
```

```shell
/usr/local/cuda/include/thrust/system/cuda/detail/bulk/detail/apply_from_tuple.hpp(118): error: no instance of function template "thrust::system::cuda::detail::merge_detail::merge_kernel::operator()" matches the argument list
            argument types are: (thrust::system::cuda::detail::bulk_::concurrent_group<thrust::system::cuda::detail::bulk_::agent<5UL>, 288UL>, const thrust::zip_iterator<thrust::tuple<thrust::counting_iterator<long long, thrust::use_default, thrust::use_default, thrust::use_default>, thrust::counting_iterator<long long, thrust::use_default, thrust::use_default, thrust::use_default>, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>>, const thrust::detail::intmax_t, const thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<long long>>, thrust::counting_iterator<long long, thrust::use_default, thrust::use_default, thrust::use_default>, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>>, const signed long, const thrust::detail::normal_iterator<thrust::pointer<int, thrust::system::cuda::detail::tag, thrust::use_default, thrust::use_default>>, const thrust::zip_iterator<thrust::tuple<thrust::discard_iterator<thrust::use_default>, thrust::discard_iterator<thrust::use_default>, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>>, const thrust::detail::compare_first<thrust::less<long long>>)
            object type is: thrust::system::cuda::detail::merge_detail::merge_kernel
          detected during:
            instantiation of "void thrust::system::cuda::detail::bulk_::detail::apply_from_tuple(Function, const thrust::tuple<Arg1, Arg2, Arg3, Arg4, Arg5, Arg6, Arg7, Arg8, thrust::null_type, thrust::null_type> &) [with Function=thrust::system::cuda::detail::merge_detail::merge_kernel, Arg1=thrust::system::cuda::detail::bulk_::concurrent_group<thrust::system::cuda::detail::bulk_::agent<5UL>, 288UL> &, Arg2=thrust::zip_iterator<thrust::tuple<thrust::counting_iterator<long long, thrust::use_default, thrust::use_default, thrust::use_default>, thrust::counting_iterator<long long, thrust::use_default, thrust::use_default, thrust::use_default>, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>>, Arg3=thrust::detail::intmax_t, Arg4=thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<thrust::device_ptr<long long>>, thrust::counting_iterator<long long, thrust::use_default, thrust::use_default, thrust::use_default>, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>>, Arg5=signed long, Arg6=thrust::detail::normal_iterator<thrust::pointer<int, thrust::system::cuda::detail::tag, thrust::use_default, thrust::use_default>>, Arg7=thrust::zip_iterator<thrust::tuple<thrust::discard_iterator<thrust::use_default>, thrust::discard_iterator<thrust::use_default>, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>>, Arg8=thrust::detail::compare_first<thrust::less<long long>>]" 
/usr/local/cuda/include/thrust/system/cuda/detail/bulk/detail/closure.hpp(50): here
```

**Comment by sdalton1** [comment]
Isolated error to counting_iterator difference_type. 

```c++
#include <thrust/device_vector.h>
#include <thrust/host_vector.h>
#include <thrust/merge.h>
#include <thrust/iterator/constant_iterator.h>
#include <thrust/iterator/counting_iterator.h>
#include <thrust/iterator/discard_iterator.h>

#include <typeinfo>
#include <iostream>

#include <type_traits>
#include <typeinfo>
#ifndef _MSC_VER
#   include <cxxabi.h>
#endif
#include <memory>
#include <string>
#include <cstdlib>

template <class T>
std::string
type_name()
{
    typedef typename std::remove_reference<T>::type TR;
    std::unique_ptr<char, void(*)(void*)> own
           (
#ifndef _MSC_VER
                abi::__cxa_demangle(typeid(TR).name(), nullptr,
                                           nullptr, nullptr),
#else
                nullptr,
#endif
                std::free
           );
    std::string r = own != nullptr ? own.get() : typeid(TR).name();
    if (std::is_const<TR>::value)
        r += " const";
    if (std::is_volatile<TR>::value)
        r += " volatile";
    if (std::is_lvalue_reference<T>::value)
        r += "&";
    else if (std::is_rvalue_reference<T>::value)
        r += "&&";
    return r;
}

int main()
{
    typedef long long index_type;
    typedef thrust::counting_iterator<index_type> Iterator;
    typedef typename Iterator::difference_type dtype;
    typedef typename Iterator::value_type vtype;
    typedef thrust::device_vector<index_type> Array;
    typedef typename Array::iterator ArrayIterator;
    typedef typename ArrayIterator::difference_type adtype;
    typedef typename ArrayIterator::value_type avtype;
    typedef thrust::constant_iterator<index_type> ConstantIterator;
    typedef typename ConstantIterator::difference_type cdtype;
    typedef typename ConstantIterator::value_type cvtype;

    dtype d1;
    vtype v1;
    adtype ad1;
    avtype av1;
    cdtype cd1;
    cvtype cv1;

    std::cout << "constant_iterator difference_type decltype(cd1) is " << type_name<decltype(cd1)>() << '\n';
    std::cout << "constant_iterator value_type      decltype(cv1) is " << type_name<decltype(cv1)>() << '\n';
    std::cout << "counting_iterator difference_type decltype( d1) is " << type_name<decltype( d1)>() << '\n';
    std::cout << "counting_iterator value_type      decltype( v1) is " << type_name<decltype( v1)>() << '\n';
    std::cout << "vector iterator   difference_type decltype(ad1) is " << type_name<decltype(ad1)>() << '\n';
    std::cout << "vector iterator   value_type      decltype(av1) is " << type_name<decltype(av1)>() << '\n';

    return 0;
}
```

```shell
constant_iterator difference_type decltype(cd1) is long
constant_iterator value_type      decltype(cv1) is long long
counting_iterator difference_type decltype( d1) is long long
counting_iterator value_type      decltype( v1) is long long
vector iterator   difference_type decltype(ad1) is long
vector iterator   value_type      decltype(av1) is long long
```

Should lines 62 (numeric_difference) and 63 (identity_) be reversed in the thrust/iterator/detail/counting_iterator_base.inl file? 

```c++
  // unlike Boost, we explicitly use std::ptrdiff_t as the difference type
  // for floating point counting_iterators
  typedef typename thrust::detail::ia_dflt_help<
    Difference,
    thrust::detail::eval_if<
      thrust::detail::is_numeric<Incrementable>::value,
        thrust::detail::eval_if<
          thrust::detail::is_integral<Incrementable>::value,
          thrust::detail::numeric_difference<Incrementable>,
          thrust::detail::identity_<std::ptrdiff_t>
        >,
      thrust::iterator_difference<Incrementable>
    >
  >::type difference;
```

Based on the comment
```
if (is_numeric){
   if (is_integral){
      then use std::ptrdiff_t // current logic returns Incrementable, right?
   }else{
      use Incrementable
   }
}else{
   use Incrementable
}
```

**Comment by jaredhoberock** [comment]
Not sure -- what do you expect the type to be? Try adding an assert to your reproducer to indicate what you expected.

**Comment by sdalton1** [comment]
Sorry the file is [counting_iterator.inl](https://github.com/thrust/thrust/blob/master/thrust/iterator/detail/counting_iterator.inl#L62) without the base part.

Here's a truly minimal version of the issue.

```c++
#include <thrust/device_vector.h>
#include <thrust/iterator/counting_iterator.h>

template<typename Size>
void my_func(const Size first_diff, const Size second_diff){}

int main()
{
    typedef long long index_type;
    thrust::device_vector<index_type> vec;
    my_func(vec.end() - vec.begin(), thrust::counting_iterator<index_type>(10) - thrust::counting_iterator<index_type>(0));
    return 0;
}
```

```shell
merge_fail.cu(89): error: no instance of function template "my_func" matches the argument list
            argument types are: (signed long, thrust::detail::intmax_t)
```
I think the vector iterator difference_type should match the counting_iterator difference_type. vec.end() - vec.begin() returns signed long (std::ptrdiff_t) while the counting_iterator returns thrust::detail::intmax_t. I believe this mismatch is causing the error in the cuda [merge](https://github.com/thrust/thrust/blob/master/thrust%2Fsystem%2Fcuda%2Fdetail%2Fmerge.inl#L191) function on line 191 when last1-first1 doesn't match last2-first2.

**Comment by jaredhoberock** [comment]
I'm still not clear about what you think is the correct behavior. Whose `difference_type` do you think is incorrect? It's not clear to me that the types in question in your program should necessarily be the same.

I'd collect this info:

1. What is the `difference_type` of `thrust::device_vector::iterator`?
2. What is the `difference_type` of `std::vector::iterator`?
3. What is the `difference_type` of `thrust::counting_iterator<long long>`?
4. What is the `difference_type` of `boost::counting_iterator<long long>`?

I expect 1. & 2. to both be `ptrdiff_t`. I'm not sure what I expect from 3. & 4.

I'd also suggest replacing `long long` with a sized integer so that we can reason about the sizes of the types in question.

**Comment by dachziegel** [comment]
`boost::counting_iterator` exhibits the same behavior as thrust. The following does not compile as well:

    #include <boost/iterator/counting_iterator.hpp>
    #include <vector>

    template<typename Size>
    void my_func(const Size first_diff, const Size second_diff){}

    int main()
    {
	    typedef long long index_type;
	    std::vector<index_type> vec;
	    my_func(vec.end() - vec.begin(), boost::counting_iterator<index_type>(10) - boost::counting_iterator<index_type>(0));
	    return 0;
    }

----

	/tmp/test.cpp:11:5: error: no matching function for call to 'my_func'
	    my_func(vec.end() - vec.begin(), boost::counting_iterator<index_type>(10) - boost::counting_iterator<index_type>(0));
	    ^~~~~~~
	/tmp/test.cpp:5:6: note: candidate template ignored: deduced conflicting types for parameter 'Size' ('long' vs. 'long long')
	void my_func(const Size first_diff, const Size second_diff){}
	     ^

However, when using `typedef int64_t index_type;` it compiles fine.

The following code provokes compiler errors to print the actual types:

    #include <boost/iterator/counting_iterator.hpp>
	#include <thrust/iterator/counting_iterator.h>
	#include <thrust/device_vector.h>
	#include <vector>
	#include <stdint.h>
	
	
	namespace type_test
	{   
	    using index_type = int64_t;
	    
	    namespace thrust
	    { 
	      namespace device_vector     { using difference_type  = ::thrust::device_vector<index_type>::iterator::difference_type; } 
	      namespace counting_iterator { using difference_type  = ::thrust::counting_iterator<index_type>::difference_type;       } 
	    }
	    namespace std
	    {
	      namespace vector           { using difference_type   = ::std::vector<index_type>::iterator::difference_type;           }        
	    }
	    namespace boost
	    { 
	      namespace counting_iterator { using difference_type  = ::boost::counting_iterator<index_type>::difference_type;       } 
	    }
	}
	
	int main()
	{
	   
	    using X1 = typename type_test::thrust::device_vector::difference_type::XXX;
	    using X2 = typename type_test::std::vector::difference_type::XXX;
	    using X3 = typename type_test::thrust::counting_iterator::difference_type::XXX;
	    using X3 = typename type_test::boost::counting_iterator::difference_type::XXX;
	    
	    return 0;
	}

---

**results:**

`using index_type = int64_t;`:
    `difference_type == long` for all

`using index_type = long long;`
    1 && 2 :  `difference_type == long`
    3 && 4 :  `difference_type == long long`

**Comment by jaredhoberock** [comment]
Thanks Manuel. This suggests to me that the root cause of the problem is elsewhere.

Probably there's some function inside of `merge` (possibly the one indicated by the original error message) that needs to be generalized to take two arguments of different types instead of two arguments of the same type.

**Comment by dachziegel** [comment]
is #655 related to this issue?

**Comment by jaredhoberock** [comment]
Fixed by #675
